staging/src/k8s.io/client-go/util/workqueue/queue.go

```diff
@@ -181,10 +185,25 @@ func (q *Type) Done(item interface{}) {
 	}
 }
 
+// isProcessing indicates if there are still items on the work queue being processed.
+// It's used to drain the work queue on an eventual shutdown
+func (q *Type) isProcessing() bool {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	return q.processing.len() != 0
+}
+
 // ShutDown will cause q to ignore all new items added to it. As soon as the
 // worker goroutines have drained the existing items in the queue, they will be
 // instructed to exit.
 func (q *Type) ShutDown() {
```

Comment:
This is a reasonable idea on the *first* sig-int.  But on the second interrupt or on sig-term we should directly shutdown directly.  Do that, I think you need to pass a `secondInterruptCh`.

Reply:
I've implemented a signal notifier which will set `forceShutdown` and interrupt the drain. I am however not sure I like it. Kubernetes controllers seems to be very opinionated when it comes to handling multiple signal interrupts, and it seems [like best practices](https://github.com/kubernetes/sample-controller/blob/master/pkg/signals/signal.go#L36-L39) even exit the process completely. Also, given that signals aren't uniform cross OS platforms I needed to add dedicated OS parts to this PR to handle this (which I am not sure the queue should / can handle).

Reply:
This package should not bake in the signal handling; that's up to the caller, there are other reasons to want to shut down.

I think what David is asking for, is something more like a second ShutDownNoWaiting method which doesn't block for the queue to clear, and ideally makes any callers blocking on ShutDown return immediately also.

Reply:
I think that makes more sense. I've implemented `ShutDownNoDrain` now which will not wait for the queue to drain when invoked, and also immediately release any callers waiting on `ShutDown`. Callers can thus implement their own signal handler and invoke `ShutDownNoDrain` in whatever situation they see fit as to force the shut down to finish immediately.

Reply:
> I haven't reviewed the tests. The interface looks reasonable to me now. Perhaps @deads2k can confirm.

I think I'm ok interface-wise, but we have now changed the contract without changing the method signature.  This call used to cause all .Gets to return as shutdown immediately and the method did not block. Now the method blocks.

This means we have to update our callers, so that in our cases (KCM, KS, KAS, etc) we have the first ctrl+c call shutdown (as we do today I think), and a second ctrl+c call a `ShutdownNoDrain`, right?

Reply:
> but we have now changed the contract without changing the method signature

I don't fully agree. We have not changed the contract, the contract said that it will drain the queue on shut down and this is not what it does today. What has changed is the implementation: which now does what the contract says. 

> This means we have to update our callers

Not necessarily. If a caller could accept a false contract, it should be able to accept a true one. Callers today usually forcefully end up doing `oc.Exit(1)` on the second ctrl+c call, this can continue being the case if callers do not want to use `ShutDownNoDrain`. The only real change this patch implements, for what concerns existing callers, is: allowing the queue to finish cleaning up after first ctrl+c and until the second ctrl+c

Reply:
> we have now changed the contract without changing the method signature

Sorry, I missed this, I thought it was net-new; I agree we should introduce a ShutdownWithDrain rather than a ShutdownNoDrain.

> If a caller could accept a false contract, it should be able to accept a true one.

Tragically, this is emphatically not true in software :( :( :(

Reply:
> I agree we should introduce a ShutdownWithDrain rather than a ShutdownNoDrain.

OK, I've updated the PR and implemented `ShutDownWithDrain`, leaving `ShutDown` as is (i.e: not blocking and not draining). It is safe to call `ShutDown` after `ShutDownWithDrain` as to force the queue to shut down immediately on the callers decision (ex: when receiving a dedicated OS signal, SIGTERM/SIGINT)